### ISSUE TITLE
httpclient with -d:ssl enabled, Valgrind reported memory leaks.

### ISSUE THREAD

**Issue by BLeAm** [nim-lang/Nim#17553]
I'm testing with httpclient and think experiencing memory leaks when turned SSL on via -d:ssl . I try to find if there's any issue related to this problem that had been reported and found #5921, according to @Yardanico 's suggestion to make a separate issue as it's a different one. So, here's my attempt.

When compiled the code below with ```nim c --gc:orc -g -d:useMalloc -d:release --panics:on main```
```nim
import httpclient

proc main =
  let client = newHttpClient()
  let res = client.request("http://www.google.com", httpMethod = HttpGet)
  echo res.code()
  client.close()

main()
```

Valgrind reported:
```
LEAK SUMMARY:
   definitely lost: 0 bytes in 0 blocks
   indirectly lost: 0 bytes in 0 blocks
     possibly lost: 56 bytes in 1 blocks
   still reachable: 615,978 bytes in 9,826 blocks
        suppressed: 0 bytes in 0 blocks
Rerun with --leak-check=full to see details of leaked memory
```

And then, If I compile the code below with ```nim c --gc:orc -g -d:useMalloc -d:release --panics:on -d:ssl main```
```nim
import httpclient

proc main =
  let client = newHttpClient()
  #aware of "s" in https
  var res = client.request("https://www.google.com", httpMethod = HttpGet)
  echo res.code()
  client.close()

main()
```
Valgrind reported:
```
LEAK SUMMARY:
   definitely lost: 367 bytes in 2 blocks
   indirectly lost: 5,778 bytes in 132 blocks
     possibly lost: 56 bytes in 1 blocks
   still reachable: 617,234 bytes in 9,843 blocks
        suppressed: 0 bytes in 0 blocks
```

I tried further with this code:
```nim
import httpclient, os

proc main =
  let url = if "-s" in commandLineParams():
    "https://www.google.com"
  else:
    "http://www.google.com"
  while true:
    let client = newHttpClient()
    var res = client.request(url, httpMethod = HttpGet)
    echo res.code()
    client.close()
    sleep(1000)

main()
```
and compiled with -d:ssl, when run the output without -s, the memory usage of the process stays at 1.3MB and don't change, on the other hand, run it with -s, the memory usage keeps climbing up over the period of testing(I once waited until it reached 10+MB then terminated the process), so I guess this's probably a hint of memory leak.

Tested on:
```
$ nim --version
Nim Compiler Version 1.5.1 [Linux: amd64]
Compiled at 2021-03-25
Copyright (c) 2006-2021 by Andreas Rumpf

git hash: 355985ac89dd13d5b317744b30842155d9321565
active boot switches: -d:release

$uname -a 
Linux iMachine 4.15.0-140-generic #144-Ubuntu SMP Fri Mar 19 14:12:35 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
```

P.S. I'm kind of in the process of learning and getting into Nim and its ecosystem. So, I might have some non-sense questions or submitted false issues, I'd like to say sorry to you guys in advance.

**Comment by ghost** [comment]
Your first two code snippets are not a bug - those 600KB are needed for the default SSL context - see https://github.com/nim-lang/Nim/blob/version-1-4/lib/pure/httpclient.nim#L320

**Comment by ghost** [comment]
That said, your last snippet doesn't leak for me either... What's your Nim version? Please follow the issue template next time, so we can easily know which Nim version you are on.

**Comment by ghost** [comment]
Hm, "definitely lost" from the second snippet might be an actual leak.

**Comment by BLeAm** [comment]
> That said, your last snippet doesn't leak for me either... What's your Nim version? Please follow the issue template next time, so we can easily know which Nim version you are on.

Thank you for your suggestion :)

I tested on 1.5.1: git hash: 355985ac89dd13d5b317744b30842155d9321565

**Comment by ghost** [comment]
After #17554 and #17558 I can confirm that there are no leaks except the expected "still reachable" ones from the default SSL context. For 15 and 35 requests the number of still reachable bytes is the same so that it's not a leak:
```
 still reachable: 610,483 bytes in 9,716 blocks
``` 
There's also
```
possibly lost: 56 bytes in 1 blocks
```
Which is also not a memory leak - it's from net.nim end of `newContext`:
```nim
    result = SslContext(context: newCTX, referencedData: initHashSet[int](),
      extraInternal: new(SslContextExtraInternal))
```
because newContext is used when creating a new default (global) SSL context. So the extraInternal never gets freed.

 Don't close this issue though, it'll automatically get closed after #17558 is merged.

**Comment by BLeAm** [comment]
@Yardanico Thank you for all the fixes and elaborations/explanations, this is so much a warm welcome, I've learned somethings via your suggestions and the way you solved the bug.
